cob_teleop/ros/src/cob_teleop.cpp

```diff
@@ -0,0 +1,611 @@
+ * \note
+ *   ROS package name: cob_teleop
+ *
+ * \author
+ *   Author: Nadia Hammoudeh García, email:nadia.hammoudeh.garcia@ipa.fhg.de
+ * \author
+ *   Supervised by: Nadia Hammoudeh García, email:nadia.hammoudeh.garcia@ipa.fhg.de
+ *
+ * \date Date of creation: August 2015
+ *
+ * \brief
+ *   Implementation of teleoperation node.
+ *
+ *****************************************************************
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *     - Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer. \n
+ *     - Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution. \n
+ *     - Neither the name of the Fraunhofer Institute for Manufacturing
+ *       Engineering and Automation (IPA) nor the names of its
+ *       contributors may be used to endorse or promote products derived from
+ *       this software without specific prior written permission. \n
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License LGPL as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License LGPL for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License LGPL along with this program.
+ * If not, see <http://www.gnu.org/licenses/>.
+ *
+ ****************************************************************/
+#include <ros/ros.h>
+
+#include <actionlib/client/simple_action_client.h>
+#include <cob_light/LightMode.h>
+#include <cob_script_server/ScriptAction.h>
+#include <cob_sound/SayAction.h>
+#include <geometry_msgs/Twist.h>
+#include <sensor_msgs/JointState.h>
+#include <sensor_msgs/Joy.h>
+#include <sensor_msgs/JoyFeedbackArray.h>
+#include <std_msgs/Float64MultiArray.h>
+#include <std_msgs/ColorRGBA.h>
+#include <std_srvs/Trigger.h>
+#include <XmlRpcValue.h>
+
+const int PUBLISH_FREQ = 100.0;
+
+class CobTeleop
+{
+public:
+
+  struct component_config{
+    std::string key;
+    std::string twist_topic_name;
+    std::string vel_group_topic_name;
+    std::string sss_default_target;
+    std::vector<double> joint_velocity;
+    ros::Publisher vel_group_controller_publisher_;
+    ros::Publisher twist_controller_publisher_;
+    std::vector<double> twist_max_vel; //max_vx_,max_vy_,max_vth_;
+    std::vector<double> twist_max_acc; //max_ax_,max_ay_,max_ath_;
```

Comment:
only base twist is planar (x, y, theta); twists for manipulators can be 6d (x, y, z, rotX, rotY, rotZ)